planning/mission_planner/src/mission_planner/mission_planner.cpp

```diff
@@ -334,19 +365,124 @@ void MissionPlanner::on_set_mrm_route(
   const SetMrmRoute::Service::Request::SharedPtr req,
   const SetMrmRoute::Service::Response::SharedPtr res)
 {
-  // TODO(Yutaka Shimizu): reroute for MRM
-  (void)req;
-  (void)res;
+  using ResponseCode = autoware_adapi_v1_msgs::srv::SetRoutePoints::Response;
+
+  if (!planner_->ready()) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The planner is not ready.");
+  }
+  if (!odometry_) {
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_UNREADY, "The vehicle pose is not received.");
+  }
+
+  const auto prev_state = state_.state;
+  change_state(RouteState::Message::CHANGING);
+
+  // Plan route.
+  const auto new_route = create_route(req);
+
+  if (new_route.segments.empty()) {
+    change_state(prev_state);
+    throw component_interface_utils::ServiceException(
+      ResponseCode::ERROR_PLANNER_FAILED, "Failed to plan a new route.");
+  }
+
+  if (!normal_route_) {
+    // if it does not set normal route, just use the new planned route
+    change_mrm_route(new_route);
+    change_state(RouteState::Message::SET);
+    res->status.success = true;
+    return;
+  }
```

Comment:
This should be moved between step1 and step2. Safety is not checked when only mrm_route is set.

Reply:
done https://github.com/autowarefoundation/autoware.universe/pull/4027/commits/d16313cb6f9ffa9065fadd59f5f30561b91f5d93